### PR TITLE
Configuration for project-local package-configurations and curations

### PR DESCRIPTION
**Pull request by MarcelBochtler**
This disables the project-local package-configuration and curations by default, as it was discussed in the [developer meeting](https://github.com/oss-review-toolkit/ort/wiki/Developer-Meeting#2021-09-02).

### CODE REVIEW

File: model/src/main/kotlin/config/AnalyzerConfiguration.kt [hunk 1]
```diff
@@ -43,5 +43,12 @@ data class AnalyzerConfiguration(
...
      */
-    val sw360Configuration: Sw360StorageConfiguration? = null
+    val sw360Configuration: Sw360StorageConfiguration? = null,
+
+    /**
+     * Enable the usage of project-local package-curations from the [RepositoryConfiguration]. If set to true apply
+     * package-curations from a local .ort.yml file. These package-curations are merged with package-curations from
+     * the command line option, while the one from .ort.yml take precedence.
+     */
+    val applyRepositoryPackageCurations: Boolean = false
```
Comment (fviernau) [inline model/src/main/kotlin/config/AnalyzerConfiguration.kt:53]: How about `enablePackageCurationsInRepositoryConfiguration`?

I believe that this option potentially applies to all components other than the analyzer.
IIRC the evaluator also has an option to re-apply curations.
So how about moving that to `OrtConfiguration`?

I presume in another PR you'll add a similar flag for `PackageConfigurations`.
These are not used by the analyzer, but evalutor / reporters. I guess these would fit into `OrtConfigurations` so
the option introduced here would fit right next to it?!

File: model/src/main/kotlin/config/EvaluatorConfiguration.kt [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties
+import com.fasterxml.jackson.annotation.JsonInclude
+
+/**
+ * The configuration model of the evaluator.
+ */
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+data class EvaluatorConfiguration(
```
Comment (fviernau) [inline model/src/main/kotlin/config/EvaluatorConfiguration.kt:30]: As proposed, move to `OrtConfiguration`...

File: cli/src/main/kotlin/commands/EvaluatorCommand.kt [hunk 3]
```diff
@@ -263,15 +264,26 @@ class EvaluatorCommand : CliktCommand(name = "evaluate", help = "Evaluate ORT re
...
 
-        val packageConfigurationProvider = SimplePackageConfigurationProvider(
-            optionPackageConfigurations + repositoryPackageConfigurations
-        )
+        if (config.evaluator.applyRepositoryPackageConfigurations) {
+            packageConfigurations += repositoryPackageConfigurations
+        } else {
+            if (repositoryPackageConfigurations.isNotEmpty()) {
+                log.warn {
+                    "Package-configurations from '$ORT_REPO_CONFIG_FILENAME' not applied because the feature is not" +
```
Comment (fviernau) [inline cli/src/main/kotlin/commands/EvaluatorCommand.kt:278]: I believe that both assumptions about the filenames are wrong.
The `.ort.yml` could also be called `my-project.ort.yml`. 
IIRC also the ort config filename can be overridden.
How about dropping the filename?

File: cli/src/main/kotlin/commands/EvaluatorCommand.kt [hunk 4]
```diff
@@ -263,15 +264,26 @@ class EvaluatorCommand : CliktCommand(name = "evaluate", help = "Evaluate ORT re
...
+            packageConfigurationOption.createProvider().getPackageConfigurations().toMutableList()
         val repositoryPackageConfigurations = ortResultInput.repository.config.packageConfigurations
-        val optionPackageConfigurations = packageConfigurationOption.createProvider().getPackageConfigurations()
 
-        val packageConfigurationProvider = SimplePackageConfigurationProvider(
-            optionPackageConfigurations + repositoryPackageConfigurations
-        )
+        if (config.evaluator.applyRepositoryPackageConfigurations) {
+            packageConfigurations += repositoryPackageConfigurations
+        } else {
```
Comment (fviernau) [inline cli/src/main/kotlin/commands/EvaluatorCommand.kt:275]: can be `else if {` for reduced nesting.

**Review by fviernau — CHANGES_REQUESTED** [review]

File: model/src/main/kotlin/config/AnalyzerConfiguration.kt (hunk 1 above)
Comment (MarcelBochtler) [reply model/src/main/kotlin/config/AnalyzerConfiguration.kt:53]: > How about `enablePackageCurationsInRepositoryConfiguration`?

I like the shorter `applyRepositoryPackageCurations` more.
However as the `.ort.yml` is called "repository configuration" in all the docs, I'll change it.

Comment (fviernau) [reply model/src/main/kotlin/config/AnalyzerConfiguration.kt]: The reason for my proposal was that I was afraid some pick-up "RepositoryPackageCuration" as new term.
Not sure if it's a valid concern.

File: cli/src/main/kotlin/commands/AnalyzerCommand.kt [hunk 5]
```diff
@@ -173,12 +174,20 @@ class AnalyzerCommand : CliktCommand(name = "analyze", help = "Determine depende
...
 
+        val packageCurations =
+            FilePackageCurationProvider.from(packageCurationsFile, packageCurationsDir).packageCurations.toMutableList()
+
+        val repositoryPackageCurations = repositoryConfiguration.curations.packages
+
+        if (config.applyRepositoryPackageCurations) {
+            packageCurations += repositoryPackageCurations
+        } else if (repositoryPackageCurations.isNotEmpty()) {
+            log.warn { "Local package-curation were not applied because the feature is not enabled." }
```
Comment (fviernau) [inline cli/src/main/kotlin/commands/AnalyzerCommand.kt:185]: `package-curation` -> `package curations`?

**Review by fviernau — CHANGES_REQUESTED** [review]

File: model/src/main/kotlin/config/AnalyzerConfiguration.kt (hunk 1 above)
Comment (fviernau) [reply model/src/main/kotlin/config/AnalyzerConfiguration.kt]: Another way to avoid that could be naming the variable `ignoreLocalPackageCurations`. What do you think?

File: cli/src/main/kotlin/commands/AnalyzerCommand.kt [hunk 6]
```diff
@@ -173,12 +174,20 @@ class AnalyzerCommand : CliktCommand(name = "analyze", help = "Determine depende
 
         val repositoryConfiguration = actualRepositoryConfigurationFile?.readValueOrNull() ?: RepositoryConfiguration()
 
+        val packageCurations =
+            FilePackageCurationProvider.from(packageCurationsFile, packageCurationsDir).packageCurations.toMutableList()
```
Comment (sschuberth) [inline cli/src/main/kotlin/commands/AnalyzerCommand.kt:178]: How about using `.toMutableSet()` here and omitting `.toSet()` in line 190 below?

File: model/src/main/kotlin/config/AnalyzerConfiguration.kt [hunk 7]
```diff
@@ -43,5 +43,5 @@ data class AnalyzerConfiguration(
     /**
      * Configuration of the SW360 package curation provider.
      */
-    val sw360Configuration: Sw360StorageConfiguration? = null
+    val sw360Configuration: Sw360StorageConfiguration? = null,
```
Comment (sschuberth) [inline model/src/main/kotlin/config/AnalyzerConfiguration.kt:46]: Trailing comma introduced.

File: model/src/main/kotlin/config/OrtConfiguration.kt [hunk 8]
```diff
@@ -48,6 +48,13 @@ data class OrtConfiguration(
      */
     val severeIssueThreshold: Severity = Severity.WARNING,
 
+    /**
+     * Enable the usage of project-local package-curations from the [RepositoryConfiguration]. If set to true apply
+     * package-curations from a local .ort.yml file. These package-curations are merged with package-curations from
+     * the command line option, while the one from .ort.yml take precedence.
+     */
+    val applyRepositoryPackageCurations: Boolean = false,
```
Comment (sschuberth) [inline model/src/main/kotlin/config/OrtConfiguration.kt:56]: Both the log statement in `AnalyzerCommand.kt` and the docs here use the term "enable". So how about also naming the property `enableRepositoryPackageCurations`?

File: cli/src/main/kotlin/commands/EvaluatorCommand.kt [hunk 9]
```diff
@@ -263,15 +263,21 @@ class EvaluatorCommand : CliktCommand(name = "evaluate", help = "Evaluate ORT re
             ortResultInput = ortResultInput.replacePackageCurations(curations)
         }
 
+        val config = globalOptionsForSubcommands.config
+
+        val packageConfigurations =
+            packageConfigurationOption.createProvider().getPackageConfigurations().toMutableList()
```
Comment (sschuberth) [inline cli/src/main/kotlin/commands/EvaluatorCommand.kt:269]: Should this also be a (mutable) set for deduplication?

File: cli/src/main/kotlin/commands/EvaluatorCommand.kt [hunk 10]
```diff
@@ -263,15 +263,21 @@ class EvaluatorCommand : CliktCommand(name = "evaluate", help = "Evaluate ORT re
...
         val repositoryPackageConfigurations = ortResultInput.repository.config.packageConfigurations
-        val optionPackageConfigurations = packageConfigurationOption.createProvider().getPackageConfigurations()
 
-        val packageConfigurationProvider = SimplePackageConfigurationProvider(
-            optionPackageConfigurations + repositoryPackageConfigurations
-        )
+        if (config.applyRepositoryPackageConfigurations) {
+            packageConfigurations += repositoryPackageConfigurations
+        } else if (repositoryPackageConfigurations.isNotEmpty()) {
+            log.warn { "Local package-configurations were not applied because the feature is not enabled." }
```
Comment (sschuberth) [inline cli/src/main/kotlin/commands/EvaluatorCommand.kt:275]: Also here: "package configurations" without a dash?

File: model/src/main/kotlin/config/OrtConfiguration.kt [hunk 11]
```diff
@@ -55,6 +55,13 @@ data class OrtConfiguration(
      */
     val applyRepositoryPackageCurations: Boolean = false,
 
+    /**
+     * Enable the usage of project-local package-configurations from the [RepositoryConfiguration]. If set to true apply
+     * package-configurations from a local .ort.yml file. These package-configurations are merged with
+     * package-configurations from the command line option, while the one from .ort.yml take precedence.
+     */
+    val applyRepositoryPackageConfigurations: Boolean = false,
```
Comment (sschuberth) [inline model/src/main/kotlin/config/OrtConfiguration.kt:63]: Also here: "enable" vs "apply"?

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: cli/src/main/kotlin/commands/AnalyzerCommand.kt (hunk 6 above)
Comment (MarcelBochtler) [reply cli/src/main/kotlin/commands/AnalyzerCommand.kt:178]: AFAIK, a Set does not guarantee to retain the order of the elements, but this is required to ensure that the local curations will take precedence over the global ones.

See Line 183:
```kt
if (config.applyRepositoryPackageCurations) {
    packageCurations += repositoryPackageCurations
} else if (repositoryPackageCurations.isNotEmpty()) {
    log.warn { "Local package-curation were not applied because the feature is not enabled." }
}
```

File: model/src/main/kotlin/config/AnalyzerConfiguration.kt (hunk 1 above)
Comment (MarcelBochtler) [reply model/src/main/kotlin/config/AnalyzerConfiguration.kt]: > Another way to avoid that could be naming the variable `ignoreLocalPackageCurations`. What do you think?

I don't like to turn around the logic when using `ignore`, because it will introduce double negative when turning this feature on.

File: cli/src/main/kotlin/commands/AnalyzerCommand.kt (hunk 6 above)
Comment (sschuberth) [reply cli/src/main/kotlin/commands/AnalyzerCommand.kt:178]: For a general set you're right, but `toMutableSet()` guarantees that "The returned set preserves the element iteration order of the original collection."

File: model/src/main/kotlin/config/AnalyzerConfiguration.kt (hunk 1 above)
Comment (MarcelBochtler) [reply model/src/main/kotlin/config/AnalyzerConfiguration.kt]: > The reason for my proposal was that I was afraid some pick-up "RepositoryPackageCuration" as new term.
> Not sure if it's a valid concern.

`LocalPackageCurations` would also introduce a new term.
I think as `Repository Configuration` is already a term used in multiple places, `repository curation` is IMO self-explanatory.

What do you think about @sschuberth's suggestion `enableRepositoryPackageCurations` [here](https://github.com/oss-review-toolkit/ort/pull/4430#discussion_r702615271)?

Comment (fviernau) [reply model/src/main/kotlin/config/AnalyzerConfiguration.kt]: Ok with me. I don't have a suggestion which solves the issue I outlined :-(

Comment (fviernau) [reply model/src/main/kotlin/config/AnalyzerConfiguration.kt]: > What do you think about @sschuberth's suggestion `enableRepositoryPackageCurations` [here](https://github.com/oss-review-toolkit/ort/pull/4430#discussion_r702615271)?

Ok with me. I don't have a suggestion which solves the issue I outlined :-(

Comment (sschuberth) [reply model/src/main/kotlin/config/AnalyzerConfiguration.kt]: PS: I was also thinking about "local" vs. "repository", and while I like "local" for being the more generic term, and ORT *used* to work fine for analyzing local non-VCS directories, that [does not really true hold anymore](https://github.com/oss-review-toolkit/ort/issues/2579#issuecomment-913166262), and we implicitly assume local directories to always be VCS working trees. Also, the term "repository" is already used e.g. in `RepositoryConfiguration`, which otherwise should also be renamed to `LocalConfiguration` or `ProjectConfiguation` (the latter actually is what I wanted to discuss in the ORT Developer Meeting for quite some time, see the [wiki](https://github.com/oss-review-toolkit/ort/wiki/Developer-Meeting)).

File: cli/src/main/kotlin/commands/AnalyzerCommand.kt (hunk 6 above)
Comment (MarcelBochtler) [reply cli/src/main/kotlin/commands/AnalyzerCommand.kt:178]: Yes in the initial transformation to a mutable map.
When adding other elements, which is being done in Line 183, the order is no longer guaranteed, right?

Comment (sschuberth) [reply cli/src/main/kotlin/commands/AnalyzerCommand.kt:178]: > When adding other elements, which is being done in Line 183, the order is no longer guaranteed, right?

Hmm, you're probably right, good point.

File: cli/src/main/kotlin/commands/AnalyzerCommand.kt [hunk 12]
```diff
@@ -56,6 +56,7 @@ import org.ossreviewtoolkit.utils.ORT_PACKAGE_CURATIONS_DIRNAME
 import org.ossreviewtoolkit.utils.ORT_PACKAGE_CURATIONS_FILENAME
 import org.ossreviewtoolkit.utils.ORT_REPO_CONFIG_FILENAME
 import org.ossreviewtoolkit.utils.expandTilde
+import org.ossreviewtoolkit.utils.log
```
Comment (sschuberth) [inline cli/src/main/kotlin/commands/AnalyzerCommand.kt:59]: The commit message still says "applyRepositoryPackageCurations".

BTW, my general recommendation to not mention newly introduced properties by name in the commit message (for exactly the reason happened here), but rather say something more general like "Introduce a configuration flag for repository package curations".

File: model/src/test/kotlin/config/OrtConfigurationTest.kt [hunk 13]
```diff
@@ -146,9 +146,7 @@ class OrtConfigurationTest : WordSpec({
                     useSsl shouldBe true
                     fromAddress shouldBe "no-reply@oss-review-toolkit.org"
                 }
-            }
```
Comment (sschuberth) [inline model/src/test/kotlin/config/OrtConfigurationTest.kt:149]: Nit: This commit is unrelated to the topics of the "surrounding" commits. While it's perfectly fine to still keep it as part of this PR, maybe just make it the last commit, so it does not "disrupt" the review flow for the "real" topic of this PR.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: model/src/main/resources/reference.conf [hunk 14]
```diff
@@ -9,6 +9,9 @@ ort {
 
   severeIssueThreshold = ERROR
 
+  enableRepositoryPackageCurations = true
```
Comment (tsteenbe) [inline model/src/main/resources/reference.conf:12]: The commit message and PR message says this feature is disable by default but in the `reference.conf` you show it as enabled.  Should we set it to false and add a comment what these flags do.

Comment (MarcelBochtler) [reply model/src/main/resources/reference.conf:12]: > The commit message and PR message says this feature is disable by default but in the `reference.conf` you show it as enabled. Should we set it to false and add a comment what these flags do.

The `reference.conf` does not reflect the default settings. See e.g. line 10 for `severeIssueThreshold`, which defaults to `WARNING`.

If the flag for `enableRepositoryPackageCurations` is not set at all, it is set to `false` and therefore disabled by default.
I did set it to `true` to check in a test if the changed value is taken into account.

> and add a comment what these flags do.

As of now the `reference.conf` does not document any of the other options.
I documented the feature in `docs/config-file-ort-yml.md`.

File: docs/config-file-ort-yml.md [hunk 15]
```diff
@@ -153,7 +153,9 @@ curations:
     concluded_license: "Apache-2.0"
  ```
 
-For findings in third-party dependencies package-configurations can be used to replace findings:
+For findings in third-party dependencies package configurations can be used to replace findings if
+`enableRepositoryPackageConfigurations` is enabled in the evaluator section of `ort.conf`, see:
```
Comment (tsteenbe) [inline docs/config-file-ort-yml.md:157]: The 'in the evaluator section' is incorrect as `enableRepositoryPackageCuration` and `enableRepositoryPackageConfigurations` both are set in the root level of `ort.conf` looking at the submitted change to `reference.conf`

Comment (tsteenbe) [reply docs/config-file-ort-yml.md:157]: I would rephrase the sentence to align it more with [config-file-package-configuration-yml.md](https://github.com/oss-review-toolkit/ort/blob/master/docs/config-file-package-configuration-yml.md) wording e.g. "To overwrite scanner findings to correct identified licenses in a dependency a package configuration can be used. Note that this feature requires `enableRepositoryPackageConfigurations` to be enabled in the `ort.conf` see [reference.conf](../model/src/main/resources/reference.conf).

File: docs/config-file-ort-yml.md [hunk 16]
```diff
@@ -178,6 +180,8 @@ The list of available options for `reason` are defined in
 Package curations can be added if you want to correct metadata of third-party dependencies.
 
 The following example corrects the source-artifact URL of the package with the id `Maven:com.example:dummy:0.0.1`.
+This requires to enable `enableRepositoryPackageCurations` in `ort.conf` like it is being done in
```
Comment (tsteenbe) [inline docs/config-file-ort-yml.md:183]: Change to: Note that this feature requires `enableRepositoryPackageCurations` to be enabled in the `ort.conf`, see [reference.conf](../model/src/main/resources/reference.conf).

File: model/src/main/kotlin/config/OrtConfiguration.kt [hunk 17]
```diff
@@ -48,6 +48,20 @@ data class OrtConfiguration(
...
+    /**
+     * Enable the usage of project-local package curations from the [RepositoryConfiguration]. If set to true apply
+     * package curations from a local .ort.yml file. These package curations are merged with package curations from
+     * the command line option, while the one from .ort.yml take precedence.
+     */
+    val enableRepositoryPackageCurations: Boolean = false,
+
+    /**
+     * Enable the usage of project-local package configurations from the [RepositoryConfiguration]. If set to true apply
+     * package configurations from a local .ort.yml file. These package configurations are merged with
```
Comment (tsteenbe) [inline model/src/main/kotlin/config/OrtConfiguration.kt:60]: Recommend to change to: If set to true, apply package configurations from a local .ort.yml file before applying those specified via the command line e.g. configurations from the.ort.yml take precedence.

File: model/src/main/kotlin/config/OrtConfiguration.kt [hunk 18]
```diff
@@ -48,6 +48,20 @@ data class OrtConfiguration(
      */
     val severeIssueThreshold: Severity = Severity.WARNING,
 
+    /**
+     * Enable the usage of project-local package curations from the [RepositoryConfiguration]. If set to true apply
+     * package curations from a local .ort.yml file. These package curations are merged with package curations from
+     * the command line option, while the one from .ort.yml take precedence.
```
Comment (tsteenbe) [inline model/src/main/kotlin/config/OrtConfiguration.kt:54]: Recommend to change to: If set to true, apply package curations from a local .ort.yml file before applying those specified via the command line e.g. curations from the.ort.yml take precedence.

**Review by tsteenbe — CHANGES_REQUESTED** [review]

File: model/src/main/kotlin/config/OrtConfiguration.kt (hunk 17 above)
Comment (MarcelBochtler) [reply model/src/main/kotlin/config/OrtConfiguration.kt:60]: Done. But replaced `e.g.` with `i.e.`.

File: model/src/main/kotlin/config/OrtConfiguration.kt (hunk 18 above)
Comment (MarcelBochtler) [reply model/src/main/kotlin/config/OrtConfiguration.kt:54]: Done. But again used `i.e.` instead of e.g. which I think is the correct term to use here.

File: docs/config-file-ort-yml.md (hunk 15 above)
Comment (MarcelBochtler) [reply docs/config-file-ort-yml.md:157]: Done. Added an `and` after findings.

File: docs/config-file-ort-yml.md [hunk 19]
```diff
@@ -153,7 +153,9 @@ curations:
     concluded_license: "Apache-2.0"
  ```
 
-For findings in third-party dependencies package-configurations can be used to replace findings:
+To overwrite scanner findings by correcting identified licenses in a dependency a package configuration can be used.
```
Comment (tsteenbe) [inline docs/config-file-ort-yml.md]: Simpler: "To correct identified licenses in a dependency you can use a package configuration to overwrite scanner findings."

File: docs/config-file-ort-yml.md [hunk 20]
```diff
@@ -153,7 +153,9 @@ curations:
     concluded_license: "Apache-2.0"
  ```
 
-For findings in third-party dependencies package-configurations can be used to replace findings:
+To correct identified licenses in a dependency you can use a package configuration to overwrite scanner findings.
```
Comment (tsteenbe) [inline docs/config-file-ort-yml.md:156]: Simpler: "To correct identified licenses in a dependency you can use a package configuration to overwrite scanner findings."

**Review by tsteenbe — CHANGES_REQUESTED** [review]

**Comment by MarcelBochtler** [comment]
@tsteenbe Do you want this merged? Otherwise I'll close this PR.

**Review by fviernau — APPROVED** [review]

**Comment by MarcelBochtler** [comment]
Adding the breaking change label, as this change by default disables repository curations / package-configurations.

**Review by sschuberth — APPROVED** [review]

**Comment by sschuberth** [comment]
Merging despite the unrelated test failures.
